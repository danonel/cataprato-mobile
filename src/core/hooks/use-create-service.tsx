import { useMutation } from 'react-query';
import { createUser } from '../features/user';
import { CreateUserDTO } from '../features/user/dto/create-user.dto';

export const useCreateUser = () => {
  const mutation = useMutation((userDetails: CreateUserDTO) => createUser().create(userDetails));

  return mutation;
};