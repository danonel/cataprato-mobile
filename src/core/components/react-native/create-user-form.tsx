import React from 'react';
import { Button, TextInput, View, Text } from 'react-native';
import { useCreateUser } from '../../hooks/use-create-service';

function CreateUserForm() {
  const [email, setEmail] = React.useState('');
  const createUserMutation = useCreateUser();

  const onSubmit = () => {
    createUserMutation.mutate({ email });
  };

  return (
    <View>
      <TextInput
        placeholder="Email"
        onChangeText={setEmail}
        value={email}
      />
      <Button onPress={onSubmit} title="Create User" />

    </View>
  );
}

export default CreateUserForm;