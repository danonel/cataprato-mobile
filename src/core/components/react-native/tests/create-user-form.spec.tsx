import React from 'react';
import { TextInput, Button } from 'react-native';
import renderer, { act } from 'react-test-renderer';
import { mocked } from 'ts-jest/utils';
import { useCreateUser } from '../../../hooks/use-create-service';
import CreateUserForm from '../create-user-form';

jest.mock('../../../hooks/use-create-service');

describe('<CreateUserForm />', () => {
  it('renders correctly', () => {
    const mockMutate = jest.fn();

    mocked(useCreateUser).mockReturnValue({
      mutate: mockMutate,
      isLoading: false,
      isError: false,
      isSuccess: false,
      error: null,
      reset: jest.fn(),
      data: null,
      isIdle: true,
    } as any);

    const tree = renderer.create(<CreateUserForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should change the state when TextInput value is changed', () => {
    const mockMutate = jest.fn();

    mocked(useCreateUser).mockReturnValue({
      mutate: mockMutate,
      isLoading: false,
      isError: false,
      isSuccess: false,
      error: null,
      reset: jest.fn(),
      data: null,
      isIdle: true,
    } as any);

    const testRenderer = renderer.create(<CreateUserForm />);
    const testInstance = testRenderer.root;

    const input = testInstance.findByType(TextInput);
    act(() => {
      input.props.onChangeText('test@email.com');
    });

    expect(input.props.value).toEqual('test@email.com');
  });

  it('should call the mutate function when the button is pressed', () => {
    const mockMutate = jest.fn();

    mocked(useCreateUser).mockReturnValue({
      mutate: mockMutate,
      isLoading: false,
      isError: false,
      isSuccess: false,
      error: null,
      reset: jest.fn(),
      data: null,
      isIdle: true,
    } as any);

    const testRenderer = renderer.create(<CreateUserForm />);
    const testInstance = testRenderer.root;

    const button = testInstance.findByType(Button);
    act(() => {
      button.props.onPress();
    });

    expect(mockMutate).toHaveBeenCalledWith({ email: '' });
  });
});