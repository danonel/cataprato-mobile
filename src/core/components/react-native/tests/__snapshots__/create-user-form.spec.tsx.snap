// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateUserForm /> renders correctly 1`] = `
<View>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Email"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Create User
      </Text>
    </View>
  </View>
</View>
`;
